rfcs/20201221-tfmot-compression-api.md

```diff
@@ -0,0 +1,447 @@
+        ```python
+        params = SVDParams(rank=32)
+        compressed_model = optimize(model, params)
+
+        loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+        compressed_model.compile(optimizer='adam', loss=loss_fn, metrics=['accuracy'])
+
+        compressed_model.fit(x_train, y_train, epochs=2)
+        compressed_model.evaluate(x_test, y_test, verbose=2)
+        ```
+    1. Deploys their compressed model to TFLite model
+        ```python
+        compressed_model.save('/tmp/model_svd_compressed')
+
+        def tflite_convert(saved_model_path, tflite_path):
+          converter = tf.lite.TFLiteConverter.from_saved_model(saved_model_path)
+          converted = converter.convert()
+          open(tflite_path, 'wb').write(converted)
+
+        tflite_convert('/tmp/model_svd_compressed',
+                       '/tmp/tflite/model_svd_compressed.tflite')
+        ```
+
+We also want to provide an example of well-known compression algorithms. Here’s algorithm list at least we have to provide:
+* [Weight clustering](https://arxiv.org/abs/1510.00149) : Most famous compression algorithm that can be used widely.
+* [WEST](https://arxiv.org/abs/1811.08417) : Example for language model area.
+* [Pruning](https://www.tensorflow.org/model_optimization/guide/pruning/pruning_with_keras) : Example for scheduling feature.
+
+### Weight compression algorithm API
+
+<p align="center">
+ <img src=20201221-tfmot-compression-api/class_graph.png />
+</p>
+
+This is an API for a layer weight based compression algorithm.
+
+First, we start from a pre-trained model which the model developer has. And then convert the pre-trained model to training phase model for compression fine-tuning training. During the convert to training phase model, We call `init_training_weights` for each tensor that we want to compress which is specified from the `get_compressible_weights` method.
+
+During the training phase, `project_training_weights` method is called for each training step. After fine-tuning training for compression is finished, we convert the training phase model to a compressed model. We only call the `compress_training_weights` function once for each compressible tensor for converting.
+
+Compressed model contains the `decompress_weights` function in the graph. It’s possible to call the `decompress_weights` for each inference step. To improve performance, we’ll cache the decompressed one depending on flags if we have enough space.
+
+```python
+class WeightCompressionAlgorithm(metaclass=abc.ABCMeta):
+  """Interface for weight compression algorithm that acts on a per-layer basis.
+
+     This allows both options of either decompressing during inference or
+     decompressing prior to inference (where compression occurs by applying a
+     tool such as zip to the model file).
+
+     This interface is a purely functional one.
+  """
+
+  @abc.abstractmethod
+  def get_compressible_weights(
```

Comment:
Why return string attributes? Why not return the actual variable?

Reply:
There's technical issues:
Currently, Wrapper layer replace the layer attribute variable by the computed tensor from the compression logic using that attribute name.

It's possible to find attribute name from the the variable. For that, we may just check all attribute names for the layer. Do you think it's more make sense?

Reply:
One point of confusion here is that it isn't clear whether the string represents the attribute name or the variable name (they're different: `layer.kernel` vs `layer.kernel.name`). Returning the actual weight object would make it clearer.

Reply:
Done. Thanks for reducing confusion.